src/bokeh/_types.pyi

```diff
@@ -0,0 +1,90 @@
+# Standard library imports
+import datetime
+from typing import Any, NotRequired, TypedDict
+
+# Bokeh imports
+from .core.property.vectorization import Expr, Field, Value
+from .models.expressions import Expression
+from .models.nodes import Node
+from .models.ranges import Factor
+from .models.text import BaseText
+from .models.transforms import Transform
+
+type NonNegative[T] = T
+type Positive[T] = T
+type Readonly[T] = T
+
+type TextLike = str | BaseText
+
+type Date = str | datetime.date
+type Datetime = str | datetime.date | datetime.datetime
+type Time = str | datetime.time
+type TimeDelta = datetime.timedelta
+
+type Color = str | tuple[int, int, int] | tuple[int, int, int, float]
+type ColorHex = Color
+type Alpha = float
+
+type FontSize = str
+
+type Angle = float
+type Percent = float
+
+type Image = Any # TODO
+
+type FieldName = str
+
+class ValueDict[ValueType](TypedDict):
+    value: ValueType
+    transform: NotRequired[Transform]
+
+class FieldDict(TypedDict):
+    field: FieldName
+    transform: NotRequired[Transform]
+
+class ExprDict(TypedDict):
+    expr: Expression
+    transform: NotRequired[Transform]
+
+type VectorInit[ValueType] = Value | Field | Expr
+type VectorDict[ValueType] = ValueDict[ValueType] | FieldDict | ExprDict
+type VectorLike[ValueType] = VectorInit[ValueType] | VectorDict[ValueType]
+
+type Vectorized[ValueType] = FieldName | ValueType | VectorLike[ValueType]
+
+type VectorizedNumber = Vectorized[float]
+
+type NumberSpec = Any
+type ColorSpec = Any
+type SizeSpec = Any
+type AngleSpec = Any
+type StringSpec = Any
+type NullStringSpec = Any
+type DistanceSpec = Any
+type NullDistanceSpec = Any
+type MarkerSpec = Any
+type DataSpec[T] = Any
+type UnitsSpec[U] = Any
+type FontStyleSpec = Any
+type TextAlignSpec = Any
+
+type TextProps = Any
+type LineProps = Any
+type FillProps = Any
+type HatchProps = Any
+type ImageProps = Any
+
+type ScalarTextProps = Any
+type ScalarLineProps = Any
+type ScalarFillProps = Any
+type ScalarHatchProps = Any
+
+type Bytes = bytes
+type JSON = str
+
+type CoordinateLike = float | Datetime | Factor
+type Coordinate = CoordinateLike | Node
+
+type CSSLength = str
+type CSSClass = str
+type CSSVariable = str
```

Comment:
Type definitions is this fill will have to be redistributed accordingly, but it was faster for now to collect everything here.